Newsletters: add 316 (2024-08-16) (#1821)
bitcoinops/bitcoinops.github.io: _topics/en/merkle-tree-vulnerabilities.md

  - title: "CVE-2017-12842: discussion about minimum transaction size"
    url: /en/newsletters/2020/05/27/#minimum-transaction-size-discussion

  - title: "CVE-2017-12842: merkle tree fix proposed in consensus cleanup soft fork"
    url:  /en/newsletters/2019/03/05/#merkle-tree-attacks

  - title: "CVE-2017-12842: discussion of lowering minimum relayable transaction size"
    url: /en/newsletters/2022/10/19/#minimum-relayable-transaction-size

  - title: "CVE-2017-12842: Bitcoin Core PR Review Club discussion about lowering min relayable tx size"
    url: /en/newsletters/2022/11/09/#bitcoin-core-pr-review-club

## Optional.  Same format as "primary_sources" above
see_also:
  - title: CVEs
    link: topic cves

## Optional.  Force the display (true) or non-display (false) of stub
## topic notice.  Default is to display if the page.content is below a
## threshold word count
#stub: false

## Required.  Use Markdown formatting.  Only one paragraph.  No links allowed.
## Should be less than 500 characters
excerpt: >
  **Merkle tree vulnerabilities** are a class of problems in the design
  of the merkle tree used by the Bitcoin consensus protocol.

---

> A merkle tree whose number of elements is a power of two can trivially

Reviewer: Maybe linking to https://en.wikipedia.org/wiki/Merkle_tree although great description here already...

be compressed into a merkle root: for a given set of elements (e.g. {A,
B, C, D}) create pairs of elements (e.g. {A,B}) and hash them together;
take the resultant internal elements (e.g., {AB, CD}}), create pairs,
and hash them together again.  Repeat until only one element remains,
the merkle root (e.g. {ABCD}).

```text
     ABCD
    /    \
  AB      CD
 /  \    /  \
A    B  C    D
```

> Merkle trees are used to allow a single 32-byte hash in each block's
> header to commit to all of the transactions in that block in a way that

Reviewer: ```suggestion
Merkle trees are used in each block's header to allow a single 32-byte hash
to commit to all of the transactions in that block in a way that
```

A
|
A
```

The subsequent sections describe several vulnerabilities associated with
this algorithm, along with their deployed mitigations.

## Invalidity caching

Bitcoin Core prevents wasting bandwidth, CPU, and memory
downloading and validating invalid blocks multiple times by caching the
header hash of recently seen blocks that are known to be invalid.
However, an attacker could modify a valid block to make it look like an
invalid block.  Consider the following valid block from our description
above:

```text
     ABCC
    /    \
  AB      CC
 /  \    /
A    B  C
```

The attacker can modify that block by duplicating the final transaction
to create a block with an identical merkle root but which is invalid (as
the second _C_ transaction spends a UTXO that was already spent in the
first _C_ transaction, the double spending of UTXOs being forbidden):

```text
     ABCC
    /    \
  AB      CC
 /  \    /  \
A    B  C    C{% comment %}<!-- skip-test -->{% endcomment %}
```

If Bitcoin Core saw the invalid form of the block first, it would store
its block header hash in its invalidity cache and refuse to accept the
valid form of the block from any of its peers until the cache was

> cleared (which occurred when the node was restarted).  An attacker could

Reviewer: ```suggestion
cleared (which occurs when the node is restarted).  An attacker could
```

use this to [eclipse attack][topic eclipse attacks] particular nodes.

The vulnerability was [reported][voight merkle] in 2012, assigned
[CVE-2012-2459][topic cves], and quietly fixed without ever being
exploited (to the best of our knowledge).

A [variation][news37 merkle] of the vulnerability was re-introduced into
Bitcoin Core in version 0.13.0 (2016) and quietly fixed in version

> 0.14.0 (2017).  Another [variation][news251 merkle] was discovered and

Reviewer: This 251 link isnt clear to me, is the PR the fix?

(edit: ok I get it now after the mention "Bitcoin Core #27608 protects against an invalidity caching vulnerability" connected the dots for me)

Author: Moved the link to the verb "fixed" to make it clear.

```text
     ABXX
    /    \
  AB      X
 /  \
A    B
```

But _X_ can also be interpreted as internal hash _CD_:

```text
     ABCD
    /    \
  AB      CD
 /  \    /  \
A    B  C    D
```

Note: that when we say "64 bytes", we are referring to a transaction's
_stripped size_, its size in the pre-segwit legacy serialization format,
which is what's used to connect a transaction to the merkle root.  This
may be different than its size in vbytes.

The Bitcoin P2P protocol message used to relay blocks, `block`, includes
a field that indicates how many transactions a block contains.  The
number of transactions in a block indicates the shape of its merkle
tree.  The derived shape of the merkle tree is used to verify that a
block header's merkle root commits to every transaction in a block.
Every committed transaction must also be valid for the block to be
valid.

The pre-image and second-preimage resistance of the SHA256d hash
function makes it extremely computationally improbable that the hash of a
valid 64-byte transaction will correspond to the two hashes of valid
internal merkle nodes.  That means it should not be possible to produce

> two valid versions of the same block with different sets of transaction:

Reviewer 2: ```suggestion
two valid versions of the same block with different sets of transactions:
```

hash to that 64-byte value.

That means Bitcoin Core's validation of merkle tree shape and
every transaction should make it impossible to trick different full
nodes into accepting different sets of transactions that can lead to a
chainsplit.

However, lightweight clients using _simplified payment verification_
(SPV) do not validate every transaction in a block (often they don't
fully validate any of them).  They simply check that a block header
merkle root commits to any transactions of interest.  In that case, it's
computationally feasible (though still expensive) to create a valid
transaction can be interpreted as a commitment one valid 32-byte
internal merkle node (the other commitment being invalid).  This attack
was discovered in 2017 and is known as [CVE-2017-12842][topic cves].

Several ways are known for mitigating or fixing this attack:

- Honest services providing merkle branch information to lightweight
  clients can tell them how many transactions the full block contains,
  allowing the lightweight client to verify that the partial merkle
  branch they're interested in corresponds to the shape of the full

>   block.  This requires the lightweight client trust the server (or poll

Reviewer 2: ```suggestion
  block.  This requires the lightweight client to trust the server (or poll
```
